File: src/geometry/wm/convert/s2.ts
import { clipLine } from '../../tools/clip.js';
import { pointFromLonLat as fromLonLat, pointToST as toST } from '../../s2/point.js';
import { fromPoint, mergeBBoxes } from '../../bbox.js';

import { geoToVector } from './vector.js';

import type {
  BBOX,
  Face,
  Feature,
  Geometry,
  MValue,
  Properties,
  S2Feature,
  STPoint,
  VectorFeature,
  VectorGeometry,
  VectorLineString,
  VectorLineStringGeometry,
  VectorMultiLineStringGeometry,
  VectorMultiPointGeometry,
  VectorMultiPolygonGeometry,
  VectorPoint,
  VectorPointGeometry,
  VectorPolygon,
  VectorPolygonGeometry,
} from '../../index.js';

/**
 * Convet a GeoJSON Feature to an S2Feature
 * @param data - GeoJSON Feature
 * @param buildBBox - optional - build a bbox for the feature if desired
 * @returns - S2Feature
 */
export function toS2<
  M = Record<string, unknown>,
  D extends MValue = Properties,
  P extends Properties = Properties,
>(
  data: Feature<M, D, P, Geometry<D>> | VectorFeature<M, D, P, VectorGeometry<D>>,
  buildBBox?: boolean,
): S2Feature<M, D, P, VectorGeometry<D>>[] {
  const { id, properties, metadata } = data;
  const res: S2Feature<M, D, P>[] = [];
  const vectorGeo =
    data.type === 'VectorFeature' ? data.geometry : geoToVector<D>(data.geometry, buildBBox);
  for (const { geometry, face } of vectorGeoToS2<D>(vectorGeo)) {
    res.push({
      id,
      type: 'S2Feature',
      face,
      properties,
      metadata,
      geometry,
    });
  }

  return res;
}

/** The resultant geometry after conversion */
export interface ConvertedGeometry<M extends MValue = Properties> {
  /** The vector geometry that was converted */
  geometry: VectorGeometry<M>;
  /** The face of the vector geometry that was converted */
  face: Face;
}
/** A list of converted geometries */
export type ConvertedGeometryList<M extends MValue = Properties> = ConvertedGeometry<M>[];

/**
 * Underlying conversion mechanic to move GeoJSON Geometry to S2Geometry
 * @param geometry - GeoJSON Geometry
 * @returns - S2Geometry
 */
export function vectorGeoToS2<M extends MValue = Properties>(
  geometry: VectorGeometry<M>,
): ConvertedGeometryList<M> {
  const { type } = geometry;
  let cGeo: ConvertedGeometryList<M>;
  if (type === 'Point') cGeo = convertGeometryPoint(geometry);
  else if (type === 'MultiPoint') cGeo = convertGeometryMultiPoint(geometry);
  else if (type === 'LineString') cGeo = convertGeometryLineString(geometry);
  else if (type === 'MultiLineString') cGeo = convertGeometryMultiLineString(geometry);
  else if (type === 'Polygon') cGeo = convertGeometryPolygon(geometry);
  else if (type === 'MultiPolygon') cGeo = convertGeometryMultiPolygon(geometry);
  else {
    throw new Error('Either the conversion is not yet supported or Invalid S2Geometry type.');
  }
  return cGeo;
}

/**
 * @param geometry - GeoJSON PointGeometry
 * @returns - S2 PointGeometry
 */
function convertGeometryPoint<M extends MValue = Properties>(
  geometry: VectorPointGeometry<M>,
): ConvertedGeometryList<M> {
  const { type, is3D, coordinates, bbox } = geometry;
  const { z, m } = coordinates;
  const [face, s, t] = toST(fromLonLat(coordinates));
  const vecBBox = fromPoint({ x: s, y: t, z });
  return [{ face, geometry: { type, is3D, coordinates: { x: s, y: t, z, m }, bbox, vecBBox } }];
}

/**
 * @param geometry - GeoJSON PointGeometry
 * @returns - S2 PointGeometry
 */
function convertGeometryMultiPoint<M extends MValue = Properties>(
  geometry: VectorMultiPointGeometry<M>,
): ConvertedGeometryList<M> {
  const { is3D, coordinates, bbox } = geometry;
  return coordinates.flatMap((coordinates) =>
    convertGeometryPoint({ type: 'Point', is3D, coordinates, bbox }),
  );
}

/**
 * @param geometry - GeoJSON LineStringGeometry
 * @returns - S2 LineStringGeometry
 */
function convertGeometryLineString<M extends MValue = Properties>(
  geometry: VectorLineStringGeometry<M>,
): ConvertedGeometryList<M> {
  const { type, is3D, coordinates, bbox } = geometry;

  return convertLineString(coordinates, false).map(({ face, line, offset, vecBBox }) => {
    return { face, geometry: { type, is3D, coordinates: line, bbox, offset, vecBBox } };
  });
}

/**
 * @param geometry - GeoJSON MultiLineStringGeometry
 * @returns - S2 MultiLineStringGeometry
 */
function convertGeometryMultiLineString<M extends MValue = Properties>(
  geometry: VectorMultiLineStringGeometry<M>,
): ConvertedGeometryList<M> {
  const { coordinates, is3D, bbox } = geometry;
  return coordinates
    .flatMap((line) => convertLineString(line, false))
    .map(({ face, line, offset, vecBBox }) => ({
      face,
      geometry: { type: 'LineString', is3D, coordinates: line, bbox, offset, vecBBox },
    }));
}

/**
 * @param geometry - GeoJSON PolygonGeometry
 * @returns - S2 PolygonGeometry
 */
function convertGeometryPolygon<M extends MValue = Properties>(
  geometry: VectorPolygonGeometry<M>,
): ConvertedGeometryList<M> {
  const { type, is3D, coordinates, bbox } = geometry;
  const res: ConvertedGeometryList<M> = [];

  // conver all lines
  const outerRing = convertLineString(coordinates[0], true);
  const innerRings = coordinates.slice(1).flatMap((line) => convertLineString(line, true));

  // for each face, build a new polygon
  for (const { face, line, offset, vecBBox: polyBBox } of outerRing) {
    const polygon: VectorPolygon<M> = [line];
    const polygonOffsets = [offset];
    for (const { face: innerFace, line: innerLine, offset: innerOffset, vecBBox } of innerRings) {
      if (innerFace === face) {
        polygon.push(innerLine);
        polygonOffsets.push(innerOffset);
        mergeBBoxes(polyBBox, vecBBox);
      }
    }

    res.push({
      face,
      geometry: {
        type,
        coordinates: polygon,
        is3D,
        bbox,
        offset: polygonOffsets,
        vecBBox: polyBBox,
      },
    });
  }

  return res;
}

/**
 * @param geometry - GeoJSON MultiPolygonGeometry
 * @returns - S2 MultiPolygonGeometry
 */
function convertGeometryMultiPolygon<M extends MValue = Properties>(
  geometry: VectorMultiPolygonGeometry<M>,
): ConvertedGeometryList<M> {
  const { is3D, coordinates, bbox, offset } = geometry;
  return coordinates.flatMap((polygon, i) =>
    convertGeometryPolygon({
      type: 'Polygon',
      is3D,
      coordinates: polygon,
      bbox,
      offset: offset?.[i],
    }),
  );
}

/** LineString converted from WM to S2 */
interface ConvertedLineString<M extends MValue = Properties> {
  face: Face;
  line: VectorLineString<M>;
  offset: number;
  vecBBox: BBOX;
}

/**
 * @param line - GeoJSON LineString
 * @param isPolygon - true if the line originates from a polygon
 * @returns - S2 LineStrings clipped to it's 0->1 coordinate system
 */
function convertLineString<M extends MValue = Properties>(
  line: VectorLineString<M>,
  isPolygon: boolean,
): ConvertedLineString<M>[] {
  const res: ConvertedLineString<M>[] = [];
  // find all the faces that exist in the line while reprojectiong
  const faces = new Set<Face>();
  // first re-project all the coordinates to S2
  const newGeometry: STPoint<M>[] = [];
  for (const { x, y, z, m } of line) {
    const [face, s, t] = toST(fromLonLat({ x, y }));
    const point: STPoint<M> = { face, s, t, z, m };
    faces.add(face);
    newGeometry.push(point);
  }
  // for each face, build a line
  for (const face of faces) {
    const line: VectorLineString<M> = [];
    for (const stPoint of newGeometry) line.push(stPointToFace(face, stPoint));
    const clippedLines = clipLine(line, [0, 0, 1, 1], isPolygon);
    for (const { line, offset, vecBBox } of clippedLines) res.push({ face, line, offset, vecBBox });
  }

  return res;
}

/**
 * @param targetFace - face you want to project to
 * @param stPoint - the point you want to project
 * @returns - the projected point
 */
function stPointToFace<M extends MValue = Properties>(
  targetFace: Face,
  stPoint: STPoint<M>,
): VectorPoint<M> {
  const { face: curFace, s, t, z, m } = stPoint;
  if (targetFace === curFace) return { x: s, y: t, z, m };

  const [rot, x, y] = FACE_RULE_SET[targetFace][curFace];
  const [newS, newT] = rotate(rot as 0 | 90 | -90, s, t);

  return { x: newS + x, y: newT + y, z, m };
}

/**
 * @param rot - rotation
 * @param s - input s
 * @param t - input t
 * @returns - new [s, t] after rotating
 */
function rotate(rot: 0 | 90 | -90, s: number, t: number): [s: number, t: number] {
  if (rot === 90) return [t, 1 - s];
  else if (rot === -90) return [1 - t, s];
  else return [s, t]; // Handles the 0° case and any other unspecified rotations
}

/**
 * Ruleset for converting an S2Point from a face to another.
 * While this this set includes opposite side faces, without axis mirroring,
 * it is not technically accurate and shouldn't be used. Instead, data should let two points travel
 * further than a full face width.
 * FACE_RULE_SET[targetFace][currentFace] = [rot, x, y]
 */
const FACE_RULE_SET: [rotation: number, moveX: number, MoveY: number][][] = [
  // Target Face 0
  [
    [0, 0, 0], // Current Face 0
    [0, 1, 0], // Current Face 1
    [90, 0, 1], // Current Face 2
    [-90, 2, 0], // Current Face 3
    [-90, -1, 0], ///  Current Face 4
    [0, 0, -1], ///  Current Face 5
  ],
  // Target Face 1
  [
    [0, -1, 0], // Current Face 0
    [0, 0, 0], // Current Face 1
    [0, 0, 1], // Current Face 2
    [-90, 1, 0], // Current Face 3
    [-90, 2, 0], // Current Face 4
    [90, 0, -1], // Current Face 5
  ],
  // Target Face 2
  [
    [-90, -1, 0], // Current Face 0
    [0, 0, -1], // Current Face 1
    [0, 0, 0], // Current Face 2
    [0, 1, 0], // Current Face 3
    [90, 0, 1], // Current Face 4
    [-90, 2, 0], // Current Face 5
  ],
  // Target Face 3
  [
    [-90, 2, 0], // Current Face 0
    [90, 0, -1], // Current Face 1
    [0, -1, 0], // Current Face 2
    [0, 0, 0], // Current Face 3
    [0, 0, 1], // Current Face 4
    [-90, 1, 0], // Current Face 5
  ],
  // Target Face 4
  [
    [90, 0, 1], // Current Face 0
    [-90, 2, 0], // Current Face 1
    [-90, -1, 0], // Current Face 2
    [0, 0, -1], // Current Face 3
    [0, 0, 0], // Current Face 4
    [0, 1, 0], // Current Face 5
  ],
  // Target Face 5
  [
    [0, 0, 1], // Current Face 0
    [-90, 1, 0], // Current Face 1
    [-90, 2, 0], // Current Face 2
    [90, 0, -1], // Current Face 3
    [0, -1, 0], // Current Face 4
    [0, 0, 0], // Current Face 5
  ],
];

File: rust/geometry/wm/convert/s2.rs
use crate::geometry::{ClipLineResultWithBBox, LonLat, S2Point, clip_line};
use alloc::{collections::BTreeSet, vec, vec::Vec};
use s2json::{
    BBox3D, Face, MValue, STPoint, VectorGeometry, VectorGeometryType, VectorLineString,
    VectorLineStringGeometry, VectorMultiLineStringGeometry, VectorMultiPointGeometry,
    VectorMultiPolygonGeometry, VectorPoint, VectorPointGeometry, VectorPolygon,
    VectorPolygonGeometry,
};

/// The resultant geometry after conversion
#[derive(Debug)]
pub struct ConvertedGeometry<M: Clone + Default = MValue> {
    /// The converted geometry
    pub geometry: VectorGeometry<M>,
    /// The face of the geometry
    pub face: Face,
}
/// A list of converted geometries
pub type ConvertedGeometryList<M> = Vec<ConvertedGeometry<M>>;

// TODO: We may be able to optimize clones. Do we just take ownership with mutables?

/// Underlying conversion mechanic to move GeoJSON Geometry to S2Geometry
pub fn convert_geometry_wm_to_s2<M: Clone + Default>(
    geometry: &VectorGeometry<M>,
) -> ConvertedGeometryList<M> {
    let mut res: ConvertedGeometryList<M> = vec![];

    match geometry {
        VectorGeometry::Point(geo) => {
            res.extend(convert_geometry_point(geo));
        }
        VectorGeometry::MultiPoint(geo) => {
            res.extend(convert_geometry_multipoint(geo));
        }
        VectorGeometry::LineString(geo) => {
            res.extend(convert_geometry_linestring(geo));
        }
        VectorGeometry::MultiLineString(geo) => {
            res.extend(convert_geometry_multilinestring(geo));
        }
        VectorGeometry::Polygon(geo) => {
            res.extend(convert_geometry_polygon(geo));
        }
        VectorGeometry::MultiPolygon(geo) => {
            res.extend(convert_geometry_multipolygon(geo));
        }
    }

    res
}

/// Convert a GeoJSON PointGeometry to a S2 PointGeometry
fn convert_geometry_point<M: Clone + Default>(
    geometry: &VectorPointGeometry<M>,
) -> ConvertedGeometryList<M> {
    let VectorPointGeometry::<M> { _type, is_3d, coordinates, bbox, .. } = geometry;
    let mut new_point = coordinates.clone();
    let ll: S2Point = (&LonLat::<M>::new(new_point.x, new_point.y, None)).into();
    let (face, s, t) = ll.to_face_st();
    new_point.x = s;
    new_point.y = t;
    let vec_bbox = Some(BBox3D::from_point(&new_point));
    vec![ConvertedGeometry {
        face: face.into(),
        geometry: VectorGeometry::Point(VectorPointGeometry {
            _type: VectorGeometryType::Point,
            coordinates: new_point,
            is_3d: *is_3d,
            bbox: *bbox,
            vec_bbox,
            ..Default::default()
        }),
    }]
}

/// Convert a GeoJSON MultiPointGeometry to S2 MultiPointGeometry
fn convert_geometry_multipoint<M: Clone + Default>(
    geometry: &VectorMultiPointGeometry<M>,
) -> ConvertedGeometryList<M> {
    let VectorMultiPointGeometry { is_3d, coordinates, bbox, .. } = geometry;
    coordinates
        .iter()
        .flat_map(|coordinates| {
            convert_geometry_point(&VectorPointGeometry {
                _type: VectorGeometryType::Point,
                is_3d: *is_3d,
                coordinates: coordinates.clone(),
                bbox: *bbox,
                ..Default::default()
            })
        })
        .collect()
}

/// Convert a GeoJSON LineStringGeometry to S2 LineStringGeometry
fn convert_geometry_linestring<M: Clone + Default>(
    geometry: &VectorLineStringGeometry<M>,
) -> ConvertedGeometryList<M> {
    let VectorLineStringGeometry { _type, is_3d, coordinates, bbox, .. } = geometry;

    convert_line_string(coordinates, false)
        .into_iter()
        .map(|cline| {
            let ConvertedLineString { face, mut line, offset, vec_bbox } = cline;
            ConvertedGeometry {
                face,
                geometry: VectorGeometry::LineString(VectorLineStringGeometry {
                    _type: VectorGeometryType::LineString,
                    is_3d: *is_3d,
                    coordinates: core::mem::take(&mut line),
                    bbox: *bbox,
                    offset: Some(offset),
                    vec_bbox: Some(vec_bbox),
                    ..Default::default()
                }),
            }
        })
        .collect()
}

/// Convert a GeoJSON MultiLineStringGeometry to S2 MultiLineStringGeometry
fn convert_geometry_multilinestring<M: Clone + Default>(
    geometry: &VectorMultiLineStringGeometry<M>,
) -> ConvertedGeometryList<M> {
    let VectorMultiLineStringGeometry { is_3d, coordinates, bbox, .. } = geometry;

    coordinates
        .iter()
        .flat_map(|line| convert_line_string(line, false))
        .map(|ConvertedLineString { face, line, offset, vec_bbox }| ConvertedGeometry {
            face,
            geometry: VectorGeometry::LineString(VectorLineStringGeometry {
                _type: VectorGeometryType::LineString,
                is_3d: *is_3d,
                coordinates: line,
                bbox: *bbox,
                offset: Some(offset),
                vec_bbox: Some(vec_bbox),
                ..Default::default()
            }),
        })
        .collect()
}

/// Convert a GeoJSON PolygonGeometry to S2 PolygonGeometry
fn convert_geometry_polygon<M: Clone + Default>(
    geometry: &VectorPolygonGeometry<M>,
) -> ConvertedGeometryList<M> {
    let VectorPolygonGeometry { _type, is_3d, coordinates, bbox, .. } = geometry;
    let mut res: ConvertedGeometryList<M> = vec![];

    // conver all lines
    let mut outer_ring = convert_line_string(&coordinates[0], true);
    let mut inner_rings = coordinates[1..].iter().flat_map(|line| convert_line_string(line, true));

    // for each face, build a new polygon
    for ConvertedLineString { face, line, offset, vec_bbox: poly_bbox } in &mut outer_ring {
        let mut polygon: VectorPolygon<M> = vec![core::mem::take(line)];
        let mut polygon_offsets = vec![*offset];
        let mut poly_bbox = *poly_bbox;
        for ConvertedLineString {
            face: inner_face,
            line: inner_line,
            offset: inner_offset,
            vec_bbox,
        } in &mut inner_rings
        {
            if inner_face == *face {
                polygon.push(inner_line);
                polygon_offsets.push(inner_offset);
                poly_bbox.merge_in_place(&vec_bbox);
            }
        }

        res.push(ConvertedGeometry {
            face: *face,
            geometry: VectorGeometry::Polygon(VectorPolygonGeometry {
                _type: VectorGeometryType::Polygon,
                is_3d: *is_3d,
                coordinates: polygon,
                bbox: *bbox,
                offset: Some(polygon_offsets),
                vec_bbox: Some(poly_bbox),
                ..Default::default()
            }),
        });
    }

    res
}

/// Convert a GeoJSON MultiPolygonGeometry to S2 MultiPolygonGeometry
fn convert_geometry_multipolygon<M: Clone + Default>(
    geometry: &VectorMultiPolygonGeometry<M>,
) -> ConvertedGeometryList<M> {
    let VectorMultiPolygonGeometry { is_3d, coordinates, bbox, offset, .. } = geometry;
    coordinates
        .iter()
        .enumerate()
        .flat_map(|(i, polygon)| {
            let offset: Option<Vec<f64>> = offset.as_ref().map(|offset| offset[i].clone());
            convert_geometry_polygon(&VectorPolygonGeometry {
                _type: VectorGeometryType::Polygon,
                is_3d: *is_3d,
                coordinates: polygon.to_vec(),
                bbox: *bbox,
                offset,
                ..Default::default()
            })
        })
        .collect()
}

/// LineString converted from WM to S2
pub struct ConvertedLineString<M: Clone + Default = MValue> {
    face: Face,
    line: VectorLineString<M>,
    offset: f64,
    vec_bbox: BBox3D,
}

/// Convert WM LineString to S2
fn convert_line_string<M: Clone + Default>(
    line: &VectorLineString<M>,
    is_polygon: bool,
) -> Vec<ConvertedLineString<M>> {
    let mut res: Vec<ConvertedLineString<M>> = vec![];
    // find all the faces that exist in the line while we re-project
    let mut faces = BTreeSet::<Face>::new();
    // first re-project all the coordinates to S2
    let mut new_geometry: Vec<STPoint<M>> = vec![];
    for VectorPoint { x: lon, y: lat, z, m, .. } in line {
        let ll: S2Point = (&LonLat::<M>::new(*lon, *lat, None)).into();
        let (face, s, t) = ll.to_face_st();
        let stpoint = STPoint { face: face.into(), s, t, z: *z, m: m.clone() };
        faces.insert(stpoint.face);
        new_geometry.push(stpoint);
    }
    // for each face, build a line
    for face in faces {
        let mut line: VectorLineString<M> = vec![];
        for st_point in &mut new_geometry {
            line.push(st_point_to_face(face, st_point));
        }
        let clipped_lines =
            clip_line(&line, BBox3D::new(0., 0., 1., 1., 0., 1.), is_polygon, None, None);
        for ClipLineResultWithBBox { line, offset, vec_bbox } in clipped_lines {
            res.push(ConvertedLineString { face, line, offset, vec_bbox });
        }
    }

    res
}

/// Given a face, rotate the point into it's 0->1 coordinate system
fn st_point_to_face<M: Clone + Default>(target_face: Face, stp: &mut STPoint<M>) -> VectorPoint<M> {
    let cur_face = stp.face;
    if target_face == cur_face {
        return VectorPoint {
            x: stp.s,
            y: stp.t,
            z: stp.z,
            m: core::mem::take(&mut stp.m),
            t: None,
        };
    }

    let (rot, x, y) = &FACE_RULE_SET[target_face as usize][cur_face as usize];
    let (new_s, new_t) = rotate(*rot, stp.s, stp.t);

    VectorPoint {
        x: new_s + *x as f64,
        y: new_t + *y as f64,
        z: stp.z,
        m: core::mem::take(&mut stp.m),
        t: None,
    }
}

/// @param rot - rotation
/// @param s - input s
/// @param t - input t
/// @returns - new [s, t] after rotating
fn rotate(rot: Rotation, s: f64, t: f64) -> (f64, f64) {
    match rot {
        Rotation::_0 => (s, t),
        Rotation::_90 => (t, 1. - s),
        Rotation::_Neg90 => (1. - t, s),
    }
}

#[derive(Debug, PartialEq, Copy, Clone)]
/// Track the rotation of a face
pub enum Rotation {
    /// No rotation
    _0,
    /// Rotate 90 degrees
    _90,
    /// Rotate -90 degrees
    _Neg90,
}

/// Ruleset for converting an S2Point from a face to another.
/// While this this set includes opposite side faces, without axis mirroring,
/// it is not technically accurate and shouldn't be used. Instead, data should let two points travel
/// further than a full face width.
/// FACE_RULE_SET[target_face][currentFace] = [rot, x, y]
pub const FACE_RULE_SET: [[(Rotation, i8, i8); 6]; 6] = [
    // Target Face 0
    [
        (Rotation::_0, 0, 0),      // Current Face 0
        (Rotation::_0, 1, 0),      // Current Face 1
        (Rotation::_90, 0, 1),     // Current Face 2
        (Rotation::_Neg90, 2, 0),  // Current Face 3
        (Rotation::_Neg90, -1, 0), //  Current Face 4
        (Rotation::_0, 0, -1),     //  Current Face 5
    ],
    // Target Face 1
    [
        (Rotation::_0, -1, 0),    // Current Face 0
        (Rotation::_0, 0, 0),     // Current Face 1
        (Rotation::_0, 0, 1),     // Current Face 2
        (Rotation::_Neg90, 1, 0), // Current Face 3
        (Rotation::_Neg90, 2, 0), // Current Face 4
        (Rotation::_90, 0, -1),   // Current Face 5
    ],
    // Target Face 2
    [
        (Rotation::_Neg90, -1, 0), // Current Face 0
        (Rotation::_0, 0, -1),     // Current Face 1
        (Rotation::_0, 0, 0),      // Current Face 2
        (Rotation::_0, 1, 0),      // Current Face 3
        (Rotation::_90, 0, 1),     // Current Face 4
        (Rotation::_Neg90, 2, 0),  // Current Face 5
    ],
    // Target Face 3
    [
        (Rotation::_Neg90, 2, 0), // Current Face 0
        (Rotation::_90, 0, -1),   // Current Face 1
        (Rotation::_0, -1, 0),    // Current Face 2
        (Rotation::_0, 0, 0),     // Current Face 3
        (Rotation::_0, 0, 1),     // Current Face 4
        (Rotation::_Neg90, 1, 0), // Current Face 5
    ],
    // Target Face 4
    [
        (Rotation::_90, 0, 1),     // Current Face 0
        (Rotation::_Neg90, 2, 0),  // Current Face 1
        (Rotation::_Neg90, -1, 0), // Current Face 2
        (Rotation::_0, 0, -1),     // Current Face 3
        (Rotation::_0, 0, 0),      // Current Face 4
        (Rotation::_0, 1, 0),      // Current Face 5
    ],
    // Target Face 5
    [
        (Rotation::_0, 0, 1),     // Current Face 0
        (Rotation::_Neg90, 1, 0), // Current Face 1
        (Rotation::_Neg90, 2, 0), // Current Face 2
        (Rotation::_90, 0, -1),   // Current Face 3
        (Rotation::_0, -1, 0),    // Current Face 4
        (Rotation::_0, 0, 0),     // Current Face 5
    ],
];
